src/analysis/processing/qgsalgorithmrescaleraster.cpp

```diff
@@ -143,9 +153,10 @@ QVariantMap QgsRescaleRasterAlgorithm::processAlgorithm( const QVariantMap &para
   int iterTop = 0;
   int iterCols = 0;
   int iterRows = 0;
-  std::unique_ptr< QgsRasterBlock > block;
-  while ( iter.readNextRasterPart( mBand, iterCols, iterRows, block, iterLeft, iterTop ) )
+  std::unique_ptr< QgsRasterBlock > inputBlock;
+  while ( iter.readNextRasterPart( mBand, iterCols, iterRows, inputBlock, iterLeft, iterTop ) )
   {
+    std::unique_ptr< QgsRasterBlock > outputBlock( new QgsRasterBlock( destProvider->dataType( 1 ), iterCols, iterRows ) );
```

Comment:
Can you explain this change? It's causing the block to be created twice...

Reply:
we need separate QgsRasterBlock for input and ouput as they might be of different data types. The issue here was that when we had a Float64 dataset, block->setValue() was written the value as a double, but the code called by destProvider->writeBlock( block.get(), mBand, iterLeft, iterTop ); assumed the block was of the data type of destProvider, that is Float32, hence chaos

Reply:
Good point, thanks!